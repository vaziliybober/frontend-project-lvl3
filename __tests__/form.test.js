/* eslint-disable no-underscore-dangle */

import nock from 'nock';
import prettier from 'prettier';
import path from 'path';
import fs, { promises as fsp } from 'fs';
import { waitFor } from '@testing-library/dom';
import '@testing-library/jest-dom';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import i18next from 'i18next';
import run from '../src/application.js';

axios.defaults.adapter = httpAdapter;
nock.disableNetConnect();

const getFixturePath = (filepath) => path.join(__dirname, '..', '__fixtures__', filepath);

const options = {
  parser: 'html',
  htmlWhitespaceSensitivity: 'ignore',
  tabWidth: 4,
};

let elements;

const prettify = (html) => prettier.format(html, options);
const getSectionTree = () => prettify(elements.section.innerHTML);

beforeEach(() => {
  const initHtml = fs.readFileSync(path.join(__dirname, '..', 'index.html')).toString();
  document.documentElement.innerHTML = initHtml;

  const section = document.querySelector('section.jumbotron');
  const form = section.querySelector('form');

  elements = {
    section,
    form,
    input: form.querySelector('input'),
    submit: form.querySelector('button'),
    feedsContainer: document.querySelector('.feeds'),
    postsContainer: document.querySelector('.posts'),
  };

  run();
});

test('form initial', async () => {
  expect(getSectionTree()).toMatchSnapshot();
});

test('form invalid url', async () => {
  elements.input.value = 'asdf';
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent(i18next.t('form.errors.invalidUrl'));
  });
  expect(getSectionTree()).toMatchSnapshot();
});

test('form invalid RSS', async () => {
  elements.input.value = 'https://ru.hexlet.io/lessons.rss';
  nock('https://cors-anywhere.herokuapp.com')
    .get('/https://ru.hexlet.io/lessons.rss')
    .reply(200, 'some invalid <pipipu> rss');
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent(i18next.t('form.errors.invalidRss'));
  });
  expect(getSectionTree()).toMatchSnapshot();
});

test('form network error', async () => {
  elements.input.value = 'https://ru.hexletttttttt.io/lessons.rss';
  nock('https://cors-anywhere.herokuapp.com')
    .get('/https://ru.hexletttttttt.io/lessons.rss')
    .replyWithError('some error');
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent(i18next.t('form.errors.networkError'));
  });
  expect(getSectionTree()).toMatchSnapshot();
});

test("form rss loaded and can't load same", async () => {
  elements.input.value = 'https://ru.hexlet.io/lessons.rss';
  const rss = await fsp.readFile(getFixturePath('rss.xml'));
  nock('https://cors-anywhere.herokuapp.com')
    .get('/https://ru.hexlet.io/lessons.rss')
    .reply(200, rss);
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent(i18next.t('form.hints.rssLoaded'));
  });
  expect(getSectionTree()).toMatchSnapshot();

  elements.input.value = 'https://ru.hexlet.io/lessons.rss';
  elements.submit.click();
  await waitFor(() => {
    expect(elements.section).toHaveTextContent(i18next.t('form.errors.repetativeUrl'));
  });
  expect(getSectionTree()).toMatchSnapshot();
});
