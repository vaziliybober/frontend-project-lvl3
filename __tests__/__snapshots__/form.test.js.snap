// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form initial 1`] = `
"<div class=\\"container-xl\\">
    <div class=\\"row\\">
        <div class=\\"col-md-10 col-lg-8 mx-auto\\">
            <h1 class=\\"display-3\\">RSS Reader</h1>
            <p class=\\"lead\\">Start reading RSS today! It is easy, it is nice.</p>
            <form>
                <div class=\\"form-row\\">
                    <div class=\\"col\\">
                        <input
                            class=\\"form-control form-control-lg\\"
                            type=\\"text\\"
                            required=\\"\\"
                            autofocus=\\"\\"
                            placeholder=\\"RSS link\\"
                            title=\\"Please, fill out this field\\"
                        />
                    </div>
                    <div class=\\"col-auto\\">
                        <button
                            class=\\"btn btn-primary btn-lg px-sm-5\\"
                            type=\\"submit\\"
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>
            <p id=\\"emailHelp\\" class=\\"form-text text-muted mb-1\\">
                Examples: https://ru.hexlet.io/lessons.rss,
                https://ru.hexlet.io/blog.rss
            </p>
        </div>
    </div>
</div>
"
`;

exports[`form invalid RSS 1`] = `
"<div class=\\"container-xl\\">
    <div class=\\"row\\">
        <div class=\\"col-md-10 col-lg-8 mx-auto\\">
            <h1 class=\\"display-3\\">RSS Reader</h1>
            <p class=\\"lead\\">Start reading RSS today! It is easy, it is nice.</p>
            <form>
                <div class=\\"form-row\\">
                    <div class=\\"col\\">
                        <input
                            class=\\"form-control form-control-lg\\"
                            type=\\"text\\"
                            required=\\"\\"
                            autofocus=\\"\\"
                            placeholder=\\"RSS link\\"
                            title=\\"Please, fill out this field\\"
                        />
                    </div>
                    <div class=\\"col-auto\\">
                        <button
                            class=\\"btn btn-primary btn-lg px-sm-5\\"
                            type=\\"submit\\"
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>
            <p id=\\"emailHelp\\" class=\\"form-text text-muted mb-1\\">
                Examples: https://ru.hexlet.io/lessons.rss,
                https://ru.hexlet.io/blog.rss
            </p>
            <div class=\\"text-danger\\">This source must contain valid RSS</div>
        </div>
    </div>
</div>
"
`;

exports[`form invalid url 1`] = `
"<div class=\\"container-xl\\">
    <div class=\\"row\\">
        <div class=\\"col-md-10 col-lg-8 mx-auto\\">
            <h1 class=\\"display-3\\">RSS Reader</h1>
            <p class=\\"lead\\">Start reading RSS today! It is easy, it is nice.</p>
            <form>
                <div class=\\"form-row\\">
                    <div class=\\"col\\">
                        <input
                            class=\\"form-control form-control-lg is-invalid\\"
                            type=\\"text\\"
                            required=\\"\\"
                            autofocus=\\"\\"
                            placeholder=\\"RSS link\\"
                            title=\\"Please, fill out this field\\"
                        />
                    </div>
                    <div class=\\"col-auto\\">
                        <button
                            class=\\"btn btn-primary btn-lg px-sm-5\\"
                            type=\\"submit\\"
                            disabled=\\"\\"
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>
            <p id=\\"emailHelp\\" class=\\"form-text text-muted mb-1\\">
                Examples: https://ru.hexlet.io/lessons.rss,
                https://ru.hexlet.io/blog.rss
            </p>
            <div class=\\"text-danger\\">This must be a valid URL</div>
        </div>
    </div>
</div>
"
`;

exports[`form network error 1`] = `
"<div class=\\"container-xl\\">
    <div class=\\"row\\">
        <div class=\\"col-md-10 col-lg-8 mx-auto\\">
            <h1 class=\\"display-3\\">RSS Reader</h1>
            <p class=\\"lead\\">Start reading RSS today! It is easy, it is nice.</p>
            <form>
                <div class=\\"form-row\\">
                    <div class=\\"col\\">
                        <input
                            class=\\"form-control form-control-lg\\"
                            type=\\"text\\"
                            required=\\"\\"
                            autofocus=\\"\\"
                            placeholder=\\"RSS link\\"
                            title=\\"Please, fill out this field\\"
                        />
                    </div>
                    <div class=\\"col-auto\\">
                        <button
                            class=\\"btn btn-primary btn-lg px-sm-5\\"
                            type=\\"submit\\"
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>
            <p id=\\"emailHelp\\" class=\\"form-text text-muted mb-1\\">
                Examples: https://ru.hexlet.io/lessons.rss,
                https://ru.hexlet.io/blog.rss
            </p>
            <div class=\\"text-danger\\">Network error</div>
        </div>
    </div>
</div>
"
`;

exports[`form rss loaded and can't load same 1`] = `
"<div class=\\"container-xl\\">
    <div class=\\"row\\">
        <div class=\\"col-md-10 col-lg-8 mx-auto\\">
            <h1 class=\\"display-3\\">RSS Reader</h1>
            <p class=\\"lead\\">Start reading RSS today! It is easy, it is nice.</p>
            <form>
                <div class=\\"form-row\\">
                    <div class=\\"col\\">
                        <input
                            class=\\"form-control form-control-lg\\"
                            type=\\"text\\"
                            required=\\"\\"
                            autofocus=\\"\\"
                            placeholder=\\"RSS link\\"
                            title=\\"Please, fill out this field\\"
                        />
                    </div>
                    <div class=\\"col-auto\\">
                        <button
                            class=\\"btn btn-primary btn-lg px-sm-5\\"
                            type=\\"submit\\"
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>
            <p id=\\"emailHelp\\" class=\\"form-text text-muted mb-1\\">
                Examples: https://ru.hexlet.io/lessons.rss,
                https://ru.hexlet.io/blog.rss
            </p>
            <div class=\\"text-success\\">RSS has been loaded</div>
        </div>
    </div>
</div>
"
`;

exports[`form rss loaded and can't load same 2`] = `
"<div class=\\"container-xl\\">
    <div class=\\"row\\">
        <div class=\\"col-md-10 col-lg-8 mx-auto\\">
            <h1 class=\\"display-3\\">RSS Reader</h1>
            <p class=\\"lead\\">Start reading RSS today! It is easy, it is nice.</p>
            <form>
                <div class=\\"form-row\\">
                    <div class=\\"col\\">
                        <input
                            class=\\"form-control form-control-lg is-invalid\\"
                            type=\\"text\\"
                            required=\\"\\"
                            autofocus=\\"\\"
                            placeholder=\\"RSS link\\"
                            title=\\"Please, fill out this field\\"
                        />
                    </div>
                    <div class=\\"col-auto\\">
                        <button
                            class=\\"btn btn-primary btn-lg px-sm-5\\"
                            type=\\"submit\\"
                            disabled=\\"\\"
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>
            <p id=\\"emailHelp\\" class=\\"form-text text-muted mb-1\\">
                Examples: https://ru.hexlet.io/lessons.rss,
                https://ru.hexlet.io/blog.rss
            </p>
            <div class=\\"text-danger\\">This URL has already been added</div>
        </div>
    </div>
</div>
"
`;
